Implement RequesterAuthorizerWithErc721 authorizers

dcroote opened the pull request:
Closes #1632 by adding two new authorizers (same chain and cross chain). In each, `RequesterAuthorizerWithErc721` corresponds to the address of the authorizer contract we deploy ([from here](https://github.com/api3dao/nft-authorizer/blob/f4bba58837f6d9d3b28254f3c2764cbe6e3bb88f/contracts/RequesterAuthorizerWithErc721.sol), incorporated completely herein along with other necessary contracts from `airnode-protocol-v1`), and `erc721s` is an array of NFTs that serve as tokens to authorize requesters.
```json
        "requesterAuthorizersWithErc721": [
          {
            "erc721s": ["0x00bDB2315678afecb367f032d93F642f64180a00"],
            "RequesterAuthorizerWithErc721": "0x999DB2315678afecb367f032d93F642f64180aa9"
          }
        ],
        "crossChainRequesterAuthorizersWithErc721": [
          {
            "erc721s": ["0x3FbDB2315678afecb367f032d93F642f64180aa6"],
            "chainType": "evm",
            "chainId": "4",
            "contracts": {
              "RequesterAuthorizerWithErc721": "0x6bbbb2315678afecb367f032d93F642f64180aa4"
            },
            "chainProvider": {
              "url": "http://127.0.0.2"
            }
          }
        ]
```

@bbenligiray  I'm still struggling a bit on the e2e test, see the second commit

Siegrift commented:
I've made the e2e test work (hacky way) in https://github.com/api3dao/airnode/commit/950f5ec9c7144472c0487247334d1aa8b8e9a7ba

The proper implementation requires a small refactor.

bbenligiray commented:
:+1: for the config format

dcroote commented:
@Siegrift  thanks for such a huge help, I really appreciate it.

> I've made the e2e test work

Unfortunately due to a legacy subtlety in `authorizer.requesterEndpointAuthorizers` empty array behavior, your request was being authorized by this authorizer instead of the new `requesterAuthorizersWithErc721` authorizer. See my last commit (in which I added a check for all authorizers being empty). I also changed dependencies, see my first commit after yours.

For [isAuthorized() within this new authorizer](https://github.com/api3dao/airnode/blob/beb4484d849e6296e170e5b5f190b20014323935/packages/airnode-protocol/contracts/dev/RequesterAuthorizerWithErc721.sol#L499-L520), it appears an NFT (token) has to be deposited... Still working through understanding that

Siegrift commented:
> For [isAuthorized() within this new authorizer](https://github.com/api3dao/airnode/blob/beb4484d849e6296e170e5b5f190b20014323935/packages/airnode-protocol/contracts/dev/RequesterAuthorizerWithErc721.sol#L499-L520), it appears an NFT (token) has to be deposited... Still working through understanding that

Yeah, sorry. I mainly focused on how to make the deployments (NFT and the respective authorizer) work. I thought that was the challenging part. Anyway, I think know how to make it work with token deposit. I'll try to fix the test today.

Siegrift commented:
@dcroote Take a look at https://github.com/api3dao/airnode/commit/b3db6aa8fcee91689eba1c4f93135674ab9265cc

Siegrift commented on packages/airnode-node/src/evm/authorization/authorization-application.test.ts:
```diff
@@ -33,7 +33,12 @@ describe('mergeAuthorizations', () => {
+    expect(logs).toEqual([
+      {
+        level: 'DEBUG',
+        message: 'Requester:requesterAddress is authorized to access Endpoint ID:endpointId for Request ID:0xapiCallId',
```
It would also be nice debug level info which requester authorized the given call.

dcroote replied:
We previously had no logging when a request was authorized, so even this is an upgrade 😅. I looked into adding the originating authorizer to the log, but that is not trivial to do based on how authorization fetching currently occurs. I think it's doable, but outside the scope of this PR. Happy to open a separate issue though.

dcroote replied:
https://github.com/api3dao/airnode/issues/1679

Siegrift commented on packages/airnode-node/src/evm/authorization/authorization-fetching.test.ts:
```diff
@@ -39,26 +41,6 @@ describe('fetch (authorizations)', () => {
     expect(res).toEqual({});
   });
 
-  it('returns true for all pending requests if authorizers arrays are empty', async () => {
```
This is removed because the logic short-circuits the function call right?

dcroote replied:
Correct

Siegrift commented on packages/airnode-node/src/evm/handlers/initialize-provider.ts:
```diff
@@ -10,30 +11,37 @@ import * as templates from '../templates';
-  return { id: 'authorizations', data: res, logs };
+  if (isEmpty(currentState.settings.authorizers.requesterEndpointAuthorizers)) {
+    return { id: 'authorizations', data: {}, logs: [] };
+  } else {
```
rm else branch to avoid nesting.

dcroote replied:
Good catch, thanks.

Siegrift commented on packages/airnode-node/src/evm/handlers/initialize-provider.ts:
```diff
@@ -142,32 +194,58 @@ export async function initializeProvider(
-  ];
+
+  // If all authorizers arrays are empty then all requests are authorized
+  const allAuthorizersEmpty = Object.values(state5.settings.authorizers).every((arr) => isEmpty(arr));
```
This should deserve an INFO log I think.

dcroote replied:
Good idea, I added the log to the if/else block below

Siegrift commented on packages/airnode-protocol/contracts/authorizers/mock/MockRequesterAuthorizerWithErc721.sol:
```diff
@@ -0,0 +1,16 @@
+
+import "../../dev/RequesterAuthorizerWithErc721.sol";
+
+contract MockRequesterAuthorizerWithErc721 is RequesterAuthorizerWithErc721 {
```
The `MockRequesterAuthorizerWithErc721` is not needed. You may just use `RequesterAuthorizerWithErc721` directly. I called it mock since it was only used for testing, but you may as well just use the real thing.

dcroote replied:
Agreed- deleted.

Siegrift commented on packages/airnode-protocol/contracts/authorizers/mock/MockRequesterAuthorizerWithErc721.sol:
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import "../../dev/RequesterAuthorizerWithErc721.sol";
```
I wonder if we want to copy the contents to the Airnode repo or use it as a dependency. Dependency soundsw better to me.

dcroote replied:
Burak suggested copy the contents:https://api3workspace.slack.com/archives/C0436H2BJKE/p1677057747622019?thread_ts=1676612819.041809&cid=C0436H2BJKE

bbenligiray replied:
I can temporarily publish https://github.com/api3dao/nft-authorizer separately if you want, but to me these feel like equivalent solutions

dcroote replied:
I was convinced by the argument that these will eventually be in this repo anyway.

Siegrift replied:
Yeah, if it's going to be moved I am ok with it. An issue in the Airnode repo would be nice though.

bbenligiray replied:
https://github.com/api3dao/airnode/issues/1675

Siegrift commented on packages/airnode-validator/src/config/config.ts:
```diff
@@ -153,23 +153,51 @@ export const chainOptionsSchema = z
   })
   .strict();
 
+// Authorizations
```
I am not sure what this comment want to say.

dcroote replied:
This comment (and the below similar ones) were meant as code dividers but I realize they're confusing so I'll remove.

Siegrift commented on packages/airnode-validator/src/config/config.ts:
```diff
@@ -153,23 +153,51 @@ export const chainOptionsSchema = z
   requesterEndpointAuthorizations: z.record(endpointIdSchema, z.array(evmAddressSchema)),
 });
 
+// Authorizers: requesterEndpointAuthorizers
```
I am not sure what this comment want to say.

Siegrift commented on packages/airnode-validator/src/config/config.ts:
```diff
@@ -153,23 +153,51 @@ export const chainOptionsSchema = z
+  chainProvider: providerSchema,
+});
+
+// Authorizers: requesterAuthorizersWithErc721
```
I am not sure what this comment want to say.

Siegrift reviewed (approved):
👍 LGTM
